src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,715 @@
+  return Rewriter::rewrite(n);
+}
+
+static Integer get_bv_const_value(Node n)
+{
+  Assert(is_bv_const(n));
+  return get_bv_const(n).getConst<BitVector>().getValue();
+}
+
+/* Note: getMinBwExpr assumes that 'expr' is rewritten.
+ *
+ * If not, all operators that are removed via rewriting (e.g., ror, rol, ...)
+ * will be handled via the default case, which is not incorrect but also not
+ * necessarily the minimum. */
+unsigned BVGaussElim::getMinBwExpr(Node expr)
+{
+  stack<Node> visit;
+  /* Maps visited nodes to the determined minimum bit-width required. */
+  unordered_map<Node, unsigned, NodeHashFunction> visited;
+  unordered_map<Node, unsigned, NodeHashFunction>::iterator it;
+
+  /* Rewrite const expr, overflows in consts are irrelevant. */
+  if (is_bv_const(expr)) { expr = get_bv_const(expr); }
+
+  visit.push(expr);
+  while (!visit.empty())
+  {
+    Node n = visit.top();
+    visit.pop();
+    it = visited.find(n);
+    if (it == visited.end())
+    {
+      visited[n] = 0;
+      visit.push(n);
+      for (const Node &nn : n)
+      {
+        visit.push(nn);
+      }
+    }
+    else if (it->second == 0)
+    {
+      /* Rewrite const expr, overflows in consts are irrelevant. */
+      if (is_bv_const(n)) { n = get_bv_const(n); }
+
+      Kind k = n.getKind();
+
+      switch (k)
+      {
+        case kind::CONST_BITVECTOR:
+        {
+          visited[n] = n.getConst<BitVector>().getValue().length();
+          break;
+        }
+
+        case kind::BITVECTOR_EXTRACT:
+        {
+          unsigned w = utils::getSize(n);
+          visited[n] = std::min(w, visited[n[0]]);
+          Assert(visited[n] <= visited[n[0]]);
+          break;
+        }
+
+        case kind::BITVECTOR_ZERO_EXTEND:
+        {
+          visited[n] = visited[n[0]];
+          break;
+        }
+
+        case kind::BITVECTOR_MULT:
+        {
+          unsigned w = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+          {
+            w += visited[n[i]];
+          }
+          if (w > utils::getSize(n)) { return 0; } /* overflow */
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_CONCAT:
+        {
+          unsigned i, wnz, nc;
+          for (i = 0, wnz = 0, nc = n.getNumChildren() - 1; i < nc; ++i)
+          {
+            unsigned wni = utils::getSize(n[i]);
+            if (n[i] != utils::mkZero(wni)) { break; }
+            /* sum of all bit-widths of leading zero concats */
+            wnz += wni;
+          }
+          /* Do not consider leading zero concats, i.e.,
+           * min bw of current concat is determined as
+           *   min bw of first non-zero term
+           *   plus actual bw of all subsequent terms */
+          visited[n] = utils::getSize(n)
+                       + visited[n[i]] - utils::getSize(n[i])
+                       - wnz;
+          break;
+        }
+
+        case kind::BITVECTOR_UREM_TOTAL:
+        case kind::BITVECTOR_SHL:
```

Comment:
Just realized that this is actually wrong. Left shifts can (and likely will) increase the BW.